src/Components/Browser.JS/src/Boot.Server.ts

```diff
@@ -2,46 +2,77 @@ import '@dotnet/jsinterop';
-import { internalFunctions as uriHelperFunctions } from './Services/UriHelper';
-import { renderBatch } from './Rendering/Renderer';
 import { fetchBootConfigAsync, loadEmbeddedResourcesAsync } from './BootCommon';
 import { CircuitHandler } from './Platform/Circuits/CircuitHandler';
 import { AutoReconnectCircuitHandler } from './Platform/Circuits/AutoReconnectCircuitHandler';
+import CircuitRegistry from './Platform/Circuits/CircuitRegistry';
+import RenderQueue, { BatchStatus } from './Platform/Circuits/RenderQueue';
+import { ConsoleLogger } from './Platform/Logging/Loggers';
+import { LogLevel, ILogger } from './Platform/Logging/ILogger';
 
-async function boot() {
-  const circuitHandlers: CircuitHandler[] = [ new AutoReconnectCircuitHandler() ];
+async function boot(): Promise<void> {
+
+  // For development.
+  // Simply put a break point here and modify the log level during
+  // development to get traces.
+  // In the future we will allow for users to configure this.
+  const logger = new ConsoleLogger(LogLevel.Error);
+
+  logger.log(LogLevel.Information, 'Booting blazor.');
+
+  const circuitHandlers: CircuitHandler[] = [new AutoReconnectCircuitHandler(logger)];
   window['Blazor'].circuitHandlers = circuitHandlers;
 
   // In the background, start loading the boot config and any embedded resources
   const embeddedResourcesPromise = fetchBootConfigAsync().then(bootConfig => {
     return loadEmbeddedResourcesAsync(bootConfig);
   });
 
-  const initialConnection = await initializeConnection(circuitHandlers);
+  const initialConnection = await initializeConnection(circuitHandlers, logger); // eslint-disable-line @typescript-eslint/no-use-before-define
+
+  const circuits = CircuitRegistry.discoverPrerenderedCircuits(document, logger);
+  for (let i = 0; i < circuits.length; i++) {
+    const circuit = circuits[i];
+    circuit.initialize();
+  }
 
   // Ensure any embedded resources have been loaded before starting the app
   await embeddedResourcesPromise;
-  const circuitId = await initialConnection.invoke<string>(
-    'StartCircuit',
-    uriHelperFunctions.getLocationHref(),
-    uriHelperFunctions.getBaseURI()
-  );
-
-  window['Blazor'].reconnect = async () => {
-    const reconnection = await initializeConnection(circuitHandlers);
-    if (!(await reconnection.invoke<Boolean>('ConnectCircuit', circuitId))) {
+
+  const startCircuit = await CircuitRegistry.startCircuit(initialConnection);
+
+  if (!startCircuit) {
+    logger.log(LogLevel.Information, 'No preregistered components to render.');
+  }
+
+  const reconnect = async (): Promise<boolean> => {
+    const reconnection = await initializeConnection(circuitHandlers, logger); // eslint-disable-line @typescript-eslint/no-use-before-define
+    const results = await Promise.all(circuits.map(circuit => circuit.reconnect(reconnection)));
+
+    if (reconnectionFailed(results)) { // eslint-disable-line @typescript-eslint/no-use-before-define
```

Comment:
If tslint gets mad about calling module-level functions before they are defined, perhaps we should disable that rule. It doesn't look useful, and these suppressions are noisy.

Reply:
I can disable the rule at the file level or for the entire project. I only had it because its on the recommended profile, but if we are ok lets turn it off for the entire project.

Reply:
Turned off the rule